agent/s3.go

```diff
@@ -2,75 +2,109 @@ package agent
 
 import (
 	"errors"
+	"fmt"
 	"os"
-	"time"
 
-	"github.com/AdRoll/goamz/aws"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/credentials/ec2rolecreds"
+	"github.com/aws/aws-sdk-go/aws/endpoints"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/buildkite/agent/logger"
 )
 
-func awsS3Auth() (aws.Auth, error) {
-	// First try to authenticate using the BUILDKITE_ ENV variables
-	buildkiteAuth, buildkiteErr := buildkiteS3EnvAuth()
-	if buildkiteErr == nil {
-		return buildkiteAuth, nil
-	}
-
-	// Passing blank values here instructs the AWS library to look at the
-	// current instances meta data for the security credentials.
-	awsAuth, awsErr := aws.GetAuth("", "", "", time.Time{})
-	if awsErr == nil {
-		return awsAuth, nil
-	}
-
-	var err error
-
-	// If they attempted to use the BUILDKITE_ ENV variables, return them
-	// that error, otherwise default to the error from AWS
-	if buildkiteErr != nil && buildkiteAuth.AccessKey != "" || buildkiteAuth.SecretKey != "" {
-		err = buildkiteErr
-	} else {
-		err = awsErr
-	}
-
-	return aws.Auth{}, err
+type credentialsProvider struct {
+	retrieved bool
 }
 
-func buildkiteS3EnvAuth() (auth aws.Auth, err error) {
-	auth.AccessKey = os.Getenv("BUILDKITE_S3_ACCESS_KEY_ID")
-	if auth.AccessKey == "" {
-		auth.AccessKey = os.Getenv("BUILDKITE_S3_ACCESS_KEY")
+func (e *credentialsProvider) Retrieve() (creds credentials.Value, err error) {
+	e.retrieved = false
+
+	creds.AccessKeyID = os.Getenv("BUILDKITE_S3_ACCESS_KEY_ID")
+	if creds.AccessKeyID == "" {
+		creds.AccessKeyID = os.Getenv("BUILDKITE_S3_ACCESS_KEY")
 	}
 
-	auth.SecretKey = os.Getenv("BUILDKITE_S3_SECRET_ACCESS_KEY")
-	if auth.SecretKey == "" {
-		auth.SecretKey = os.Getenv("BUILDKITE_S3_SECRET_KEY")
+	creds.SecretAccessKey = os.Getenv("BUILDKITE_S3_SECRET_ACCESS_KEY")
+	if creds.SecretAccessKey == "" {
+		creds.SecretAccessKey = os.Getenv("BUILDKITE_S3_SECRET_KEY")
 	}
 
-	if auth.AccessKey == "" {
+	if creds.AccessKeyID == "" {
 		err = errors.New("BUILDKITE_S3_ACCESS_KEY_ID or BUILDKITE_S3_ACCESS_KEY not found in environment")
 	}
-	if auth.SecretKey == "" {
+	if creds.SecretAccessKey == "" {
 		err = errors.New("BUILDKITE_S3_SECRET_ACCESS_KEY or BUILDKITE_S3_SECRET_KEY not found in environment")
 	}
 
+	e.retrieved = true
 	return
 }
 
-func awsS3Region() (region aws.Region, err error) {
+func (e *credentialsProvider) IsExpired() bool {
+	return !e.retrieved
+}
+
+func awsS3Credentials() *credentials.Credentials {
```

Comment:
Does this support the same ENV vars as this?

https://github.com/goamz/goamz/blob/master/aws/aws.go#L340